### PR TITLE
[FIX] web_editor, website: prevent history system to break public widget

### PR DESCRIPTION
**Pull request by Guillaume-gdi**
TODO

- [ ] When an attribute change, restart public widget [see this comment](https://github.com/odoo/odoo/pull/134979#discussion_r1491052676)

### tl;dr:

Before this commit, undo and redo actions could break the public widgets
because they were not destroyed or started according to the DOM changes
made by the history system. This commit permits to start and destroy
the public widgets according to the DOM changes made by the history
system.

This commit fixes the issue and adds a test.

### Long story:

[This commit] has introduced a new editor and a new history management
system. The previous system consisted of saving the entire editable
element at each history step. Thanks to [this other commit], all public
widgets in the editable element were destroyed and restarted with
`content_will_be_destroyed` and `content_was_recreated`. To prevent
duplicate content, [an original content backup system] had to be added
(`o_temp_auto_element`) for blocks with dynamic content.

The new history management system introduced with the new editor doesn't
save all editable content at every step. It just retains modifications
between each history step. Unfortunately, the principle of destroying
and restarting public widgets has not been implemented in the new
history system, which is causing problems. This commit restores this
mechanism, taking advantage of the new history system which allows us to
destroy and restart only those public widgets that are concerned by the
event (undo/redo) requested (only start on element added and only
destroy on element removed).

Despite the changes of this commit, there are still problems with the
history system and public widgets (which are active in edit mode). For
example, public widgets must disconnect the history system observer when
they need to modify their target. In some cases, this isn't done
correctly, but luckily it doesn't always cause problems thanks to the
timeout set on the history observer. In addition, some public widgets
create HTML elements when they start, but don't remove them when they're
destroyed.

Given that this is a more or less significant behavior change , we've
decided to merge this into master to ensure that it doesn't cause any
regression in stable versions. This change may be backported to earlier
versions later.

### Steps to showcase the issue:

- Edit a website page
- Add the "Chart" snippet
- Undo
- Redo
- Try to use the options of the chart

=> An error can randomly occur because the public widget hasn't been
destroyed.

[This commit]: https://github.com/odoo/odoo/commit/499e8860f194a7eb09a2e224b713a362923d77ac
[this other commit]: https://github.com/odoo/odoo/commit/a153ed42a09f8b7f5e0865112eb7d5affc22a353
[an original content backup system]: https://github.com/odoo/odoo/commit/d1a867553da2c99f2895e1ea9ed29f5d6b295a1f

task-3605429

### CODE REVIEW

File: addons/website/static/src/snippets/s_instagram_page/000.js [hunk 1]
```diff
@@ -73,16 +73,6 @@ const InstagramPage = publicWidget.Widget.extend({
...
-        if (!iframeEl) {
-            // TODO: fix this case. We should never end up here. It happens when
-            // - Drop Instagram snippet
-            // - Undo
-            // - Drop a new Instagram snippet
-            // => The listener of the first one is still active because the
-            // public widget has not been destroyed.
-            window.removeEventListener("message", this.__onMessage);
-            return;
-        }
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_instagram_page/000.js:85]: I take the opportunity to remove this.

Note that if you want to check that my test is failing without the fix, you will have to apply the changes I made in this public widget.

File: addons/website/static/tests/tours/public_widget_history.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    },
+    {
+        content: "Wait for the instagram page to be removed",
+        trigger: "iframe body:not(:has(.s_instagram_page))",
+        run: () => {
+            // In addition, we wait for few milliseconds to be sure that the
+            // public widget is well destroyed.
+            setTimeout(() => {
+                document.body.classList.add("after_destroy");
+            }, 500);
```
Comment (Guillaume-gdi) [inline addons/website/static/tests/tours/public_widget_history.js:29]: I know this kind of things is not ideal but I can't make the test running as expected (and failing when expected) without using this hack (two times actually) :grimacing:

**Review by Guillaume-gdi — COMMENTED** [review]
@qsm-odoo This PR is ready for your review.

Note that I finally did the tour with the Instagram snippet because chart is not causing errors every time due to the history timeout. The Instagram block is always failing so it seems better.

I also hesitated to add a new public widget in test_website only for this test... if you prefer I can do it. The test would probably be easier / less hacky (I guess).

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 3]
```diff
@@ -270,6 +270,8 @@ export class OdooEditor extends EventTarget {
                 // `convertNumericToUnit` as odoo-editor utils to avoid this
                 getCSSVariableValue: () => null,
                 convertNumericToUnit: x => x,
+                beforeHistorySystemRemoveElement: () => {},
+                afterHistorySystemAddElement: () => {},
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:274]: Either or `beforeHistorySystemRemove**s**Element` or `beforeHistorySystemElementRemoval`

File: addons/website/static/tests/tours/public_widget_history.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        },
+    },
+    {
+        content: "Ensure we wait 500ms",
+        trigger: "body.after_destroy",
+        run: () => {}, // It's a check.
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/public_widget_history.js:35]: If you target master, should not it be `isCheck: true` ?

File: addons/website/static/tests/tours/public_widget_history.js (hunk 2 above)
Comment (qsm-odoo) [reply addons/website/static/tests/tours/public_widget_history.js:29]: Yeah, we should avoid that, especially since you increase the test tour by 1 second for no good reason. Maybe you should indeed create a specific widget for this test, you would not need to do that. But it does not need to be in test_website: this is only for tests that required python things (like routes) or records we don't want to be installed for test purpose only. But you can declare a public widget only added in a test bundle in the website app. Also using the instagram snippet may not be right because it makes external call (to instagram), so we cannot use it in a normal test.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Seeing the fix, it seems quite safe. Maybe we can still target master, let it be used a few weeks and if nothing bad happens, consider backporting it.

**Comment by Guillaume-gdi** [comment]
@qsm-odoo It's adapted :+1:

File: addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js [hunk 5]
```diff
@@ -379,9 +379,35 @@ export class WysiwygAdapterComponent extends Wysiwyg {
...
+        });
+    }
+    /**
+     * Starts the widgets after an history undo/redo adds a block.
+     *
+     * @private
+     * @param {HTMLElement} el - the element that has been added
+     */
+    _afterHistorySystemAddsElement(el) {
+        this._websiteRootEvent("widgets_start_request", {
```
Comment (qsm-odoo) [inline addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js:406]: Starting widgets is a async operation. When widgets are started due to an option change, we probably queue that async operation in the editor mutex. Should not it be done here too?

File: addons/website/static/tests/tours/public_widget_history.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        isCheck: true,
+    },
+]);
+
+
+odoo.loader.bus.addEventListener("module-started", (e) => {
+    if (e.detail.moduleName !== "@website/../tests/tours/public_widget_history"
+        && odoo.loader.modules.get("@web/legacy/js/public/public_widget")) {
+        const publicWidget =
+            odoo.loader.modules.get("@web/legacy/js/public/public_widget")[Symbol.for("default")];
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/public_widget_history.js:39]: What's this? Why can't you just do a normal with that public widget definition?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
I'd like a bit more info and history in the commit message:
- mentioning the "content_will_be_destroy / something like" system in 14.0 (with relevant commits)
- mentioning the temp-auto-content commit in that version too
- Why they were removed (new editor)
- Remaining issues known due to history
- Why this is merged in master and not 15.0 (tricky fix that could be backported later) (in the end: is it that tricky?)

Thanks :+1:

File: addons/website/static/tests/tours/public_widget_history.js (hunk 6 above)
Comment (Guillaume-gdi) [reply addons/website/static/tests/tours/public_widget_history.js:39]: I think it's the way to go since [this other commit](https://github.com/odoo/odoo/commit/e0fdb6a0b4bfaf9c04b52c1f74a266192d0d0676)

File: addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js (hunk 5 above)
Comment (Guillaume-gdi) [reply addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js]: @qsm-odoo I did this change :arrow_up: and it is better IMO ! 

The problem is that history repeats itself, as I introduce the change now (which do what [this old commit](https://github.com/odoo/odoo/commit/a153ed42a09f8b7f5e0865112eb7d5affc22a353) did more or less) I introduce the same issue. And so we need something more to fix as you did in [this commit](https://github.com/odoo/odoo/commit/d1a867553da2c99f2895e1ea9ed29f5d6b295a1f) in the past. I think with something similar to what you wanted to do in [this PR](https://github.com/odoo/odoo/pull/74923), we can solve the issue and the the redo/undo flows will be robust.

Just wanted your opinion before moving forward :)
